vendor/github.com/robert-milan/go-object-interning/object_intern.go

```diff
@@ -0,0 +1,484 @@
+package goi
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	gos "github.com/replay/go-generic-object-store"
+	"github.com/tmthrgd/shoco"
+)
+
+// ObjectIntern stores a map of uintptrs to interned objects.
+// The string key is itself using an interned object for its data pointer
+type ObjectIntern struct {
+	sync.RWMutex
+	conf       *ObjectInternConfig
+	Store      gos.ObjectStore
+	ObjIndex   map[string]uintptr
+	compress   func(in []byte) []byte
+	decompress func(in []byte) ([]byte, error)
+}
+
+// NewObjectIntern returns a new ObjectIntern. Pass in
+// nil if you want to use the default configuration, otherwise
+// pass in a custom configuration
+func NewObjectIntern(c *ObjectInternConfig) *ObjectIntern {
+	oi := &ObjectIntern{
+		conf:     Config,
+		Store:    gos.NewObjectStore(100),
+		ObjIndex: make(map[string]uintptr),
+	}
+	if c != nil {
+		oi.conf = c
+	}
+
+	// set compression and decompression functions
+	switch oi.conf.CompressionType {
+	case SHOCO:
+		oi.compress = func(in []byte) []byte {
+			return shoco.Compress(in)
+		}
+		oi.decompress = func(in []byte) ([]byte, error) {
+			b, err := shoco.Decompress(in)
+			return b, err
+		}
+	default:
+		oi.compress = func(in []byte) []byte {
+			return in
+		}
+		oi.decompress = func(in []byte) ([]byte, error) {
+			return in, nil
+		}
+	}
+
+	return oi
+}
+
+// CompressionFunc returns the current compression func used by the library
+func (oi *ObjectIntern) CompressionFunc() func(in []byte) []byte {
+	return oi.compress
+}
+
+// DecompressionFunc returns the current decompression func used by the library
+func (oi *ObjectIntern) DecompressionFunc() func(in []byte) ([]byte, error) {
+	return oi.decompress
+}
+
+// Compress returns a compressed version of in as a []byte
+// It is important to keep in mind that not all values can be compressed,
+// so this may at times return the original value
+func (oi *ObjectIntern) Compress(in []byte) []byte {
+	return oi.compress(in)
+}
+
+// Decompress returns a decompressed version of in as a []byte and nil on success.
+// On failure it returns an error
+func (oi *ObjectIntern) Decompress(in []byte) ([]byte, error) {
+	return oi.decompress(in)
+}
+
+// CompressSz returns a compressed version of in as a string
+// It is important to keep in mind that not all values can be compressed,
+// so this may at times return the original value
+func (oi *ObjectIntern) CompressSz(in string) string {
```

Comment:
what does `Sz` mean? I think in the standard lib they just add `String` to a function to signify taking in strings, so we can follow that